include/circt/Support/UnusedOpPruner.h

```diff
@@ -0,0 +1,72 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_UNUSEDOPPRUNER_H
+#define CIRCT_SUPPORT_UNUSEDOPPRUNER_H
+
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+
+namespace circt {
+
+/// Utility that tracks operations that have potentially become unused and
+/// allows them to be cleaned up at a later point. Useful to make passes clean
+/// up dead code.
+struct UnusedOpPruner {
+  /// Mark an op the be erased later if it is unused at that point.
+  void eraseLaterIfUnused(Operation *op) {
+    if (op)
```

Comment:
nit: Maybe `assert(op)` instead of null guard?

Reply:
Yeah that's a good idea :+1: